src/libsyntax/parse/lexer/tokentrees.rs

```diff
@@ -203,21 +205,26 @@ impl<'a> TokenTreesReader<'a> {
             },
             _ => {
                 let tt = TokenTree::Token(self.token.take());
-                // Note that testing for joint-ness here is done via the raw
-                // source span as the joint-ness is a property of the raw source
-                // rather than wanting to take `override_span` into account.
-                // Additionally, we actually check if the *next* pair of tokens
-                // is joint, but this is equivalent to checking the current pair.
-                let raw = self.string_reader.peek_span_src_raw;
                 self.real_token();
-                let is_joint = raw.hi() == self.string_reader.peek_span_src_raw.lo()
-                    && self.token.is_op();
+                let is_joint = self.joint_to_prev == Joint && self.token.is_op();
```

Comment:
`self.token.is_op()` makes me think that perhaps we should check that the previous token is also an `op`?

That is, we currently can say `Join` for an identifier followed by `(`, for example.

Reply:
All the observable behavior is represented by the [`Spacing`](https://doc.rust-lang.org/proc_macro/enum.Spacing.html) enum returned by [`Punct::spacing`](https://doc.rust-lang.org/proc_macro/struct.Punct.html#method.spacing).

It has pretty specific documentation - we are "revealing" the jointness knowledge only for `Punct + Punct` pairs and additionally for `Punct(') + Ident`.
Any other jointness is not revealed, `Ident + (` in particular is not revealed because `(` is not `is_op`, so the implementation looks correct.

Perhaps we should do this hiding entirely at the proc macro interface border though, and keep the full knowledge internally.

Reply:
>in particular is not revealed because ( is not is_op, so the implementation looks correct.

Ah, indeed, it needs to be `Ident-`, for example.

Yeah, I think we should either handle this completely on proc-macro layer, or completely in the `TokenTreesReader`, but this is unrelated to PR at hand